docs: Add user guide for Node Remove procedure in Bluetooth mesh (#9056)
nrfconnect/sdk-nrf: doc/nrf/ug_bt_mesh_node_removal.rst

.. _ug_bt_mesh_node_removal:

Removing a node from a mesh network
###################################

.. contents::
   :local:
   :depth: 2

In certain circumstances, there can be a need to remove (unprovision) a node from Bluetooth® mesh network.

> Though the unprovisioning the node is simple, there are some additional steps need to be taken in order to keep security of the network.

Author: Included as a part of Mia's suggestion.

This page describes how to do this properly.

The procedure consists of two steps:

1. Removing the node

> 2. Updating a network key on the rest of the nodes.

Author: With removing `The procedure consists of two steps:` part?

I'm uncertain if all steps should be added here. I  see this as just a highlight of the procedure.

> 2. Updating a network key on the rest of the nodes.

Author: The table of content is anyway present at the right top corner

> 2. Updating a network key on the rest of the nodes.

Reviewer: Yes, I thought we could skip the sentence "The procedure consists of two steps:" and go to the list straight from "Though the unprovisioning of the node is a straightforward procedure, some additional steps need to be taken to maintain the security of the network:"
And the thought was also to give the user complete list of these additional steps straight away.
But this is just a suggestion, and it is not wrong the way you write it either.

> The procedure consists of two steps:
> 
> 1. Removing the node
> 2. Updating a network key on the rest of the nodes.

Reviewer: But, if we are not changing the structure, we need to do the following changes:

Change "Though the unprovisioning of the node is a straightforward procedure, some additional steps need to be taken to maintain the security of the network:" to "Though the unprovisioning of the node is a straightforward procedure, some additional steps need to be taken to maintain the security of the network."

In addition to
```suggestion
The procedure consists of two steps:

1. Removing the node.
2. Updating the network key on the rest of the nodes.
```

> 2. Updating a network key on the rest of the nodes.

Reviewer: I see your point :)

Removing the node
*****************

To remove the node from the network and reset its configuration, send the Config Node Reset message from a node running the Configuration Client model, which is usually instantiated on a provisioner.

When using |NCS| API:

1. Call the :c:func:`bt_mesh_cfg_cli_node_reset` function to send Node Reset message to the node to be reset.
2. Find the node in CDB using :c:func:`bt_mesh_cdb_node_get`.
3. Remove the node from CDB using :c:func:`bt_mesh_cdb_node_del`.

When using the nRF Mesh mobile app as a provisioner:

1. On the Network screen, tap the mesh device node to be reset.
2. Scroll down and find :guilabel:`Reset Node` section.
3. Press on :guilabel:`RESET` button.

> Updating a network key

Reviewer: ```suggestion
Updating the network key
```

**********************

Even if the Config Node Reset message has been sent to the node and the Config Node Reset Status message was received, this doesn't guarantee that the node has removed all configuration data.
To ensure that the removed node is not able to decrypt network messages after being reset, update the network key of the subnet the removed node belonged to.

This procedure can be used in any other cases when a network key needs to be updated.
For example, when there is a suspicion that the network credentials have been compromised.
The same procedure can be used to update an application key.

.. note::
   Updating keys is currently not supported by the nRF Mesh mobile app.

The procedure of updating a network (application) key is described below.

Uploading a new key
===================

Upload a new network key (or application key) to the nodes by sending the Config NetKey (AppKey) Update message to every node including the node with the Configuration Client model.

> The snippet below demonstrates how to do this using |NCS| API:

Reviewer: ```suggestion
The snippet below demonstrates how to do this using the |NCS| API:
```

.. code-block:: c

    static uint8_t send_net_key_update(struct bt_mesh_cdb_node *node, void *user_data)
    {
        struct bt_mesh_cdb_subnet *subnet = user_data;
        uint8_t status = 0;
        int err;

>         err = bt_mesh_cfg_cli_net_key_update(node->net_idx, node->addr, subnet->net_idx,

Reviewer 2: ```suggestion
        err = bt_mesh_cfg_net_key_update(node->net_idx, node->addr, subnet->net_idx,
```
Guess this applies other places as well.

Author: No, without `_cli_` suffix is an old API which will eventually be removed. The new API has `_cli_` suffix. It should be added after upmerge. That is why `DNM` added.

Reviewer 2: Ahh... my bad 👍

                                             subnet->keys[1].net_key, &status);
        if (err || status) {
            printk("Unable to update NetKey on %#x (err %d, status %u)\n", node->addr, err,
                   status);
        } else {
            printk("Updated NetKey with index %u on %#x\n", subnet->net_idx, node->addr);
        }

        return BT_MESH_CDB_ITER_CONTINUE;
    }

    static int update_net_key(uint16_t net_idx, uint8_t *new_net_key)
    {
        struct bt_mesh_cdb_subnet *subnet;

        subnet = bt_mesh_cdb_subnet_get(net_idx);
        if (!subnet) {
            printk("Unable to get subnet\n");
            return -EINVAL;
        }

        /* Store the new network key in CDB. */
        memcpy(subnet->keys[1].net_key, new_net_key, 16);
        bt_mesh_cdb_subnet_store(subnet);

        /* Send the new network key to each node. */
        bt_mesh_cdb_node_foreach(send_net_key_update, subnet);
        return 0;
    }

    ...

    uint8_t new_net_key[16] = { ... };
    update_net_key(BT_MESH_NET_PRIMARY, new_net_key);

Switching to the new key
========================

> After uploading the new key, command the nodes to use it when sending messages. To do that, change the Key Refresh Phase to 2 by sending the Config Key Refresh Phase Set message with Transition field set to 0x2.

Reviewer: ```suggestion
After uploading the new key, configure the nodes to use it when sending messages.
To do that, change the Key Refresh Phase to 2 by sending the Config Key Refresh Phase Set message with Transition field set to 0x2.
```

> Send the message to at least one node. The rest nodes will switch the Key Refresh Phase through Secure Network Beacon. But, in this case, it may take a while for all nodes to eventually switch the phase since the time between 2 consecutive Secure Network Beacons is ~10 seconds. You can speed this process up by sending the Config Key Refresh Phase Set message to all nodes that needs to be updated.
> 
> After switching the Key Refresh Phase to 2, a node will decrypt messages using both, old and new keys, but encrypt only using the new key. Therefore, when updating a network key, make sure to update the phase first on the node running the Configuration Client model so that it is able to decrypt the Config Key Refresh Phase Status message from other nodes when they switch to the new network key.
> 
> The snippet below demonstrates how to set the Key Refresh Phase on all nodes using |NCS| API:

Reviewer: ```suggestion
Send the message to at least one node.
The rest of the nodes will switch the Key Refresh Phase through Secure Network Beacon.
But in this case, it may take a while for all nodes to eventually switch the phase since the time between 2 consecutive Secure Network Beacons is ~10 seconds.
You can speed up this process by sending the Config Key Refresh Phase Set message to all nodes that need to be updated.

After switching the Key Refresh Phase to 2, a node will decrypt messages using both old and new keys, but encrypt only using the new key.
Therefore, when updating a network key, make sure to update the phase on the node running the Configuration Client model first, so it is able to decrypt the Config Key Refresh Phase Status message from other nodes when they switch to the new network key.

The snippet below demonstrates how to set the Key Refresh Phase on all nodes using the |NCS| API:
```

    {
        struct bt_mesh_cdb_subnet *subnet = user_data;
        uint8_t phase;
        uint8_t status;
        int err;

        err = bt_mesh_cfg_cli_krp_set(node->net_idx, node->addr, subnet->net_idx, subnet->kr_phase,
                                      &status, &phase);
        if (err || status) {
            printk("Unable to set Key Refresh Phase on %#x (err %d, status %u)\n",
                   node->addr, err, status);
        } else {
            printk("Set Key Refresh Phase to %u on %#x\n", phase, node->addr);
        }

        return BT_MESH_CDB_ITER_CONTINUE;
    }

    static int update_key_refresh_phase(uint8_t net_idx, uint8_t phase)
    {
        struct bt_mesh_cdb_subnet *subnet;

        subnet = bt_mesh_cdb_subnet_get(net_idx);
        if (!subnet) {
            printk("Unable to get subnet\n");
            return -EINVAL;
        }

        subnet->kr_phase = phase;
        bt_mesh_cdb_subnet_store(subnet);

        bt_mesh_cdb_node_foreach(send_key_refresh_phase_set, subnet);
        return 0;
    }

    ...

    update_key_refresh_phase(BT_MESH_NET_PRIMARY, BT_MESH_KR_PHASE_2);

> If the Key Refresh Phase to be changed through Secure Network Beacons, wait until all nodes have changed the Key Refresh Phase to 2. This can be done by sending the Config Key Refresh Phase Get message to a specific node. To retrieve the Key Refresh Phase from a node using |NCS| API, use :c:func:`bt_mesh_cfg_cli_krp_get`.

Reviewer 2: ```suggestion
If the Key Refresh Phase is suposed to be changed through Secure Network Beacons, wait until all nodes have changed the Key Refresh Phase to 2. This can be done by sending the Config Key Refresh Phase Get message to a specific node. To retrieve the Key Refresh Phase from a node using |NCS| API, use :c:func:`bt_mesh_cfg_cli_krp_get`.
```

Author: Chose Mia's suggestion with just `is`.

> If the Key Refresh Phase to be changed through Secure Network Beacons, wait until all nodes have changed the Key Refresh Phase to 2. This can be done by sending the Config Key Refresh Phase Get message to a specific node. To retrieve the Key Refresh Phase from a node using |NCS| API, use :c:func:`bt_mesh_cfg_cli_krp_get`.

Reviewer: ```suggestion
If the Key Refresh Phase is to be changed through Secure Network Beacons, wait until all nodes have changed the Key Refresh Phase to 2.
This can be done by sending the Config Key Refresh Phase Get message to a specific node.
To retrieve the Key Refresh Phase from a node using the |NCS| API, use the :c:func:`bt_mesh_cfg_cli_krp_get` function.
```

Removing the old key
====================

When all nodes are in the Key Refresh Phase 2, the old key needs to be removed. To do that, switch the Key Refresh Phase to 3 by sending the Config Key Refresh Phase Set message with the Transition field set to 0x3.

> The same logic as for the phase 2 applies here. Either send this message to one of nodes (not necessary the node with the Configuration Client model) and wait while other nodes receive Secure Network Beacon, or send the message to each node.

Author: Included as a part of Mia's suggestion.

> When all nodes are in the Key Refresh Phase 2, the old key needs to be removed. To do that, switch the Key Refresh Phase to 3 by sending the Config Key Refresh Phase Set message with the Transition field set to 0x3.
> 
> The same logic as for the phase 2 applies here. Either send this message to one of nodes (not necessary the node with the Configuration Client model) and wait while other nodes receive Secure Network Beacon, or send the message to each node.
> 
> The snippet below demonstrates how to send the message to each node this using |NCS| API:

Reviewer: ```suggestion
When all nodes are in the Key Refresh Phase 2, the old key needs to be removed.
To do that, switch the Key Refresh Phase to 3 by sending the Config Key Refresh Phase Set message with the Transition field set to 0x3.

The same logic as for the phase 2 applies here.
Either send this message to one of the nodes (not necessarily the node with the Configuration Client model) and wait while other nodes receive Secure Network Beacon, or send the message to each node.

The snippet below demonstrates how to send the message to each node using the |NCS| API:
```

.. code-block:: c

    static int swap_net_keys_in_cdb(uint8_t net_idx)
    {
        struct bt_mesh_cdb_subnet *subnet;

        subnet = bt_mesh_cdb_subnet_get(net_idx);
        if (!subnet) {
            printk("Unable to get subnet\n");
            return -EINVAL;
        }

        memcpy(subnet->keys[0].net_key, subnet->keys[1].net_key, 16);
        memset(subnet->keys[1].net_key, 0, 16);
        bt_mesh_cdb_subnet_store(subnet);
    }

    ...

    update_key_refresh_phase(BT_MESH_NET_PRIMARY, BT_MESH_KR_PHASE_3);

    /* Replace the old key with the new one in CDB. */
    swap_net_keys_in_cdb(BT_MESH_NET_PRIMARY);

> Once all nodes switched the Key Refresh Phase to 3, the procedure completes.

Reviewer: ```suggestion
Once all nodes have switched the Key Refresh Phase to 3, the procedure completes.

```

To remove the node using the |NCS| API, complete the following steps:

1. Call the :c:func:`bt_mesh_cfg_cli_node_reset` function to send a Config Node Reset message to the node to be reset.
#. Find the node in the Configuration Database (CDB) using the :c:func:`bt_mesh_cdb_node_get` function.
#. Remove the node from the CDB using the :c:func:`bt_mesh_cdb_node_del` function.

When using the nRF Mesh mobile app as a provisioner, do the following:

1. On the Network screen, tap the mesh device node to be reset.
#. Scroll down and find the :guilabel:`Reset Node` section.
#. Press the :guilabel:`RESET` button.

Updating the network key
************************

Even if the Config Node Reset message has been sent to the node, and the Config Node Reset Status message was received, this doesn't guarantee that the node has removed all configuration data.
To ensure that the removed node is not able to decrypt network messages after being reset, update the network key of the subnet the removed node belonged to.

This procedure can be used in any other cases when a network key needs to be updated.
For example, in cases where there is a suspicion of network credentials being compromised.

> The same procedure can be used to update an application key.

Reviewer 3: All AppKeys are associated with a network key. They always form a pair. 
> The added AppKey can be used by the node only as a pair with the specified NetKey.

> The same procedure can be used to update an application key.

Author: Should I say how surprising it is to see the application key requirement in Config AppKey Add message?

        }

        return BT_MESH_CDB_ITER_CONTINUE;
    }

    static int update_net_key(uint16_t net_idx, uint8_t *new_net_key)
    {
        struct bt_mesh_cdb_subnet *subnet;

        subnet = bt_mesh_cdb_subnet_get(net_idx);
        if (!subnet) {
            printk("Unable to get subnet\n");
            return -EINVAL;
        }

        /* Store the new network key in CDB. */
        memcpy(subnet->keys[1].net_key, new_net_key, 16);
        bt_mesh_cdb_subnet_store(subnet);

        /* Send the new network key to each node. */
        bt_mesh_cdb_node_foreach(send_net_key_update, subnet);
        return 0;
    }

    ...

    uint8_t new_net_key[16] = { ... };
    update_net_key(BT_MESH_NET_PRIMARY, new_net_key);

Switching to the new key
========================

After uploading the new key, configure the nodes to use it when sending messages.
To do that, change the Key Refresh Phase to 2 by sending the Config Key Refresh Phase Set message with Transition field set to 0x2.

Send the message to at least one node.
The rest of the nodes will switch the Key Refresh Phase through Secure Network Beacon.

> But in this case, it may take a while for all nodes to eventually switch the phase since the time between 2 consecutive Secure Network Beacons is ~10 seconds.

Reviewer 3: ```suggestion
But in this case, it may take a while for all nodes to eventually switch the phase since the time between 2 consecutive Secure Network Beacons is ~10 seconds and some nodes can backoff from sending secure network beacons up to 600 seconds.
```

You can speed up this process by sending the Config Key Refresh Phase Set message to all nodes that need to be updated.

After switching the Key Refresh Phase to 2, a node will decrypt messages using both old and new keys, but encrypt only using the new key.
Therefore, when updating a network key, make sure to update the phase on the node running the Configuration Client model first, so it is able to decrypt the Config Key Refresh Phase Status message from other nodes when they switch to the new network key.

The snippet below demonstrates how to set the Key Refresh Phase on all nodes using the |NCS| API:

.. code-block:: c

>     static uint8_t send_key_refresh_phase_set(struct bt_mesh_cdb_node *node, void *user_data)

Reviewer 3: It almost feels like we should push this into the shell app. Then there is no need of these snippets and the whole user guide can be explained using shell commands. If someone wants to do it using code, they can then refer to the shell app which will be more complete and functioning. This will also help in avoiding the maintenance of snippets here. That way guide will become more compact.

In the intro, you could say something like:
"This user guide shows how to execute the key refresh procedure using mesh shell commands. These commands utilize |NCS| APIs and can be referred if user wants to implement the procedure by themselves ..."

Author: Yes, I generally agree that it would be the correct to move this snippet to some sample that compiles and runs. However I'm not sure if moving it to `mesh_shell` or adding to `subsys/bluetooth/mesh/shell.c` commands would be correct. `subsys/bluetooth/mesh/shell.c` is basically an interface for mesh API. Moving this snippet there will make this code as de facto a standard and people will use without adjusting it to their needs. This code is still in a snippet state as error handling is not done correctly (`bt_mesh_cdb_node_foreach` doesn't allow to pass back error code, it has no error checking between steps). It could be part of `mesh_shell`, but then not clear where to add this command (it can't be added to `mesh` command, only `mesh models` allows to add commands dynamically). And I should also wait for the next upmerge before finalizing the PR. Not sure if all these hassle makes sense here...

Reviewer 3: This will create two ways of doing things. One, potentially using shell, and one using this guide. If shell commands are not able to execute the Key Refresh procedure by invoking the individual commands, then something is missing in the config client APIs or CDB APIs, or mesh shell.

We could do this in shell afterward though and then replace (or remove) these parts from the user guide.

Author: Then I suggest to do this later.

        if (err || status) {
            printk("Unable to set Key Refresh Phase on %#x (err %d, status %u)\n",
                   node->addr, err, status);
        } else {
            printk("Set Key Refresh Phase to %u on %#x\n", phase, node->addr);
        }

        return BT_MESH_CDB_ITER_CONTINUE;
    }

    static int update_key_refresh_phase(uint8_t net_idx, uint8_t phase)
    {
        struct bt_mesh_cdb_subnet *subnet;

        subnet = bt_mesh_cdb_subnet_get(net_idx);
        if (!subnet) {
            printk("Unable to get subnet\n");
            return -EINVAL;
        }

        subnet->kr_phase = phase;
        bt_mesh_cdb_subnet_store(subnet);

        bt_mesh_cdb_node_foreach(send_key_refresh_phase_set, subnet);
        return 0;
    }

    ...

    update_key_refresh_phase(BT_MESH_NET_PRIMARY, BT_MESH_KR_PHASE_2);

If the Key Refresh Phase is to be changed through Secure Network Beacons, wait until all nodes have changed the Key Refresh Phase to 2.
This can be done by sending the Config Key Refresh Phase Get message to a specific node.
To retrieve the Key Refresh Phase from a node using the |NCS| API, use the :c:func:`bt_mesh_cfg_cli_krp_get` function.

> Removing the old key

Reviewer 3: ```suggestion
Revoking the old key
```
?

1. Call the :c:func:`bt_mesh_cfg_cli_node_reset` function to send a Config Node Reset message to the node to be reset.
#. Find the node in the Configuration Database (CDB) using the :c:func:`bt_mesh_cdb_node_get` function.
#. Remove the node from the CDB using the :c:func:`bt_mesh_cdb_node_del` function.

When using the nRF Mesh mobile app as a provisioner, do the following:

1. On the Network screen, tap the mesh device node to be reset.
#. Scroll down and find the :guilabel:`Reset Node` section.
#. Press the :guilabel:`RESET` button.

Updating the network key
************************

Even if the Config Node Reset message has been sent to the node, and the Config Node Reset Status message was received, this doesn't guarantee that the node has removed all configuration data.
To ensure that the removed node is not able to decrypt network messages after being reset, update the network key(s) that were known to the removed node.

This procedure can be used in any other cases when a network key needs to be updated.
For example, in cases where there is a suspicion of network credentials being compromised.

The same procedure can be used to update an application key.

> To update an application key, the associated network key must also be started before updating the application key.

Reviewer 3: It reads bit strange. Something is missing?
```suggestion
To update an application key, the associated network key update must also be started before starting the update for the application key.
```

Author: I think I copied your suggestion from here: https://github.com/nrfconnect/sdk-nrf/pull/9056#discussion_r1011972485
